text/0001-blinking-cursors-and-periodic-animation.md

```diff
@@ -0,0 +1,182 @@
+
+# Reference-level explanation
+
+### TL;DR
+
+ * The problem of a blinking cursor input can be generalized to _periodic animation_ support.
+ * The `iced` runtime only performs an update/draw cycle in response to application events and user interaction, but to support periodic animation, we need the runtime to perform a cycle whenever a widget needs it.
+ * The current interface (the [Widget](https://docs.rs/iced_native/0.2.2/iced_native/widget/trait.Widget.html) trait) between the runtime and widgets has no means to indicate that a widget is animating, and hence requires a re-draw at sometime in the future.
+
+_The numbering of these points aligns to that of the guided explanation in the previous section_.
+
+### 1: Event loop changes
+
+The event loop needs to be changed to track the soonest moment which an update needs to occur, and to perform an update/draw cycle at this moment.
+
+The intermediate state of a user interface (at the transition of a update/draw cycle) is stored in the `State` struct, from `native/src/program/state.rs`. A new field, `next_animation_draw: AnimationState` can be added to track the next draw required by widgets. More on the `AnimationState` type later, but just know that this type encapsulates the animation requirements of widgets, and this value can be obtained by calling a method on the root widget.
+
+Modifying the behavior of the wait state is surprisingly trivial, thanks to the underlying use of `winit`. Instead of setting the event loops' `control_flow` to `Wait`, we just set it to `WaitUntil( <time of soonest update> )`. This changes the behavior to wait until the next event, or the provided time (whichever is sooner).
+
+As the event loop only updates/draws in response to events/user-interaction, we emit a new synthetic event `AnimationTick` to drive the update cycle.
+
+### 2: Widget trait changes
+
+An update is needed to the [Widget](https://docs.rs/iced_native/0.2.2/iced_native/widget/trait.Widget.html) trait to communicate the animation requirements of a widget - in our case, that the `TextInput` widget needs to be re-drawn in 500ms.
+
+We propose adding a new method to the trait, with a default implementation that just indicate no animation is taking place:
+
+```rust
+  fn next_animation(&self) -> AnimationState {
+    AnimationState::NotAnimating
+  }
+```
+
+Widgets that need to animate (such as our `TextInput`) can implement this method to indicate an animation is required:
+
+```rust
+  fn next_animation(&self) -> AnimationState {
+    AnimationState::AnimateIn(std::time::Instant::now().checked_add(Duration::from_millis(500)))
+  }
+```
+
+Remaining consistent with the stateless and intuitive feel of the widget trait, `next_animation` is called as part of the update loop, to get the latest set of animation requirements.
```

Comment:
It may be interesting to explore how the implementation of this method will look for container widgets, such as `Column`, `Row`, or `PaneGrid`.

As you point out later, I imagine we will need to reduce all the `AnimationState`s of the children and take the smaller `Instant`, if there is any.

Reply:
Yepp exactly! The prototype used `min()`:

https://github.com/twitchyliquid64/iced/commit/22ff1f3b55d3b2041c25c7a6c20a9b3deb982f57#diff-95a8a7ef0f8ed9a198179b8d94a27082c107635e7328eb5c9168bc55be53b821R225-R231

 via the ords traits: https://github.com/twitchyliquid64/iced/commit/0376c762df1a0378ebbacc1a7f6dd7c116c6c33d#diff-0808b7a4461985abe8485a6cea5c1bef4476a0f0577f9779733669b27d9b69f1R19-R38

But we can use any kind of reduction trait for consistency.